### PR TITLE
fix lots of issues with atlas

### LINKED ISSUE #18750: atlas overwrites user files
### Example
```
cd $dir
git clone https://github.com/nim-lang/Nim
cd Nim
sh build_all.sh
ls ..
Nim
bin/nim c -r tools/atlas/atlas.nim clone https://github.com/disruptek/balls
ls ..
```

### Current Output
it generates directories outside of source tree
```
Nim             balls           grok            nim-bytes2human npeg            packages        sync            testes          ups
```
and in fact in my use case that prompted this issue, it affected and modified existing files I had in Nim/.. as that directory already contained some cloned packages.

### Expected Output
it should generate directories self-contained within source tree:
```
Nim
```
possibly under: Nim/.atlas, similar to `node_modules/`

rationale: things should be kept self contained, so you won't interfere with other local installations of nim (i have more than 1 nim folder for development purposes, branches don't solve everything).

### Additional Information
1.5.1 98f7254fb25d402a3cad8e7de81207448fc7b322

see also https://github.com/nim-lang/Nim/issues/18751

### LINKED ISSUE #18751: atlas cmd writes nim.cfg to wrong destination: cannot open: balls/nim.cfg [IOError]
### Example
```
cd $dir
git clone https://github.com/nim-lang/Nim
cd Nim
sh build_all.sh
bin/nim c -r tools/atlas/atlas.nim clone https://github.com/disruptek/balls
```

### Current Output
[Warning] (nim-bytes2human) package has no tagged releases

/private/tmp/d04/foo2/Nim/tools/atlas/atlas.nim(473) atlas
/private/tmp/d04/foo2/Nim/tools/atlas/atlas.nim(456) main
/private/tmp/d04/foo2/Nim/tools/atlas/atlas.nim(396) patchNimCfg
/private/tmp/d04/foo2/Nim/lib/system/io.nim(866) writeFile
Error: unhandled exception: cannot open: balls/nim.cfg [IOError]
Error: execution of an external program failed: '/private/tmp/d04/foo2/Nim/tools/atlas/atlas clone https://github.com/disruptek/balls'

### Expected Output
should write to $nim/config/nim.cfg, not $nim/balls/nim.cfg,

### Example 2
bin/nim c -r tools/atlas/atlas.nim clone cligen
Error: unhandled exception: cannot open: cligen/nim.cfg [IOError]

### Additional Information
1.5.1 98f7254fb25d402a3cad8e7de81207448fc7b322

### LINKED ISSUE #18753: atlas tries to checkout incorrect commit when a `requires: dep` doesn't specify a commit
after trying to fix other atlas issues (https://github.com/nim-lang/Nim/issues/18750, https://github.com/nim-lang/Nim/issues/18751) i'm running into the following issue:

### Example
after fixing https://github.com/nim-lang/Nim/issues/18750, https://github.com/nim-lang/Nim/issues/18751, running this:

bin/nim c -r tools/atlas/atlas.nim clone https://github.com/disruptek/balls

### Current Output
after instrumenting code to produce an assert instead of erroring with zero context (`quit(1)` is bad as it doesn't any context on errors; at least there should be a debug flag to allow giving errors, possibly at the nim-level so it applies to all programs calling `quit`):

/Users/timothee/git_clone/nim/Nim_prs/tools/atlas/atlas.nim(498) atlas
/Users/timothee/git_clone/nim/Nim_prs/tools/atlas/atlas.nim(479) main
/Users/timothee/git_clone/nim/Nim_prs/tools/atlas/atlas.nim(380) clone
/Users/timothee/git_clone/nim/Nim_prs/tools/atlas/atlas.nim(284) checkoutCommit
/Users/timothee/git_clone/nim/Nim_prs/tools/atlas/atlas.nim(168) warn
/Users/timothee/git_clone/nim/Nim_prs/tools/atlas/atlas.nim(164) message
/Users/timothee/git_clone/nim/Nim_prs/lib/system/assertions.nim(38) failedAssertImpl
/Users/timothee/git_clone/nim/Nim_prs/lib/system/assertions.nim(28) raiseAssert
/Users/timothee/git_clone/nim/Nim_prs/lib/system/fatal.nim(53) sysFatal

it also shows this:
[Warning] (nim-bytes2human) package has no tagged releases;

and after instrumentation:
```
 (name: ..., url: "https://github.com/juancarlospaco/nim-bytes2human", commit: "<invalid commit>", rel: strictlyLess);
```

the problem is that https://github.com/disruptek/balls recursively depends on https://github.com/juancarlospaco/nim-bytes2human which doesn't have a tagged release;

note that https://github.com/disruptek/testes/blob/d9db2ad09aa38fc26625341e1b666602959e144f/testes.nimble (which is checked out at that revision by atlas) contains:

```
requires "https://github.com/juancarlospaco/nim-bytes2human"
```

### Expected Output
it should work because nim-bytes2human doesn't contain a specific revision, so should checkout HEAD instead of attempting to checkout a (nonexistent) tag

Either that or we should clarify somewhere that `requires "https://github.com/juancarlospaco/nim-bytes2human"` is invalid, or that a tag must be specified;

in either case, better error reporting and diagnostic is needed here; even if https://github.com/juancarlospaco/nim-bytes2human gets fixed (which is easy and likely to happen), the same problem will occur in other instances.

### Additional Information
1.5.1 98f7254fb25d402a3cad8e7de81207448fc7b322

the culprit is this code:
```nim
      if tokens.len == 1:
        # nimx uses dependencies like 'requires "sdl2"'.
        # Via this hack we map them to the first tagged release.
        # (See the `isStrictlySmallerThan` logic.)
        tokens.add "<"
        tokens.add InvalidCommit
```

### PR DESCRIPTION
**Pull request by timotheecour**
* fix https://github.com/nim-lang/Nim/issues/18753
* fix https://github.com/nim-lang/Nim/issues/18750
* fix https://github.com/nim-lang/Nim/issues/18751
* adds add (for now required) flag `--atlasdir:dir` specifying root of atlas dirs; the previous hard-coded logic was flawed (refs https://github.com/nim-lang/Nim/issues/18750)
* adds a (for now required) flag specifying where to write cfg file `--outfilecfg:file`
* lots of code cleanups

## after this PR
this now works instead of failing (with several unrelated errors)
```
bin/nim r tools/atlas/atlas.nim --atlasdir:/tmp/d06e/atlas --outfilecfg:/tmp/d06e/my.nim.cfg clone https://github.com/disruptek/balls
```

note, the --path generated are correct wrt --atlasDir and --outFileCfg

## future work
- [ ] add support for `NIM_ATLAS_DIR`, analog to `NIMBLE_DIR`
- [ ] atlas should eventually be able to install a dependencies as part of bootstrapping process (`sh build_all.sh`), not just for tools/compiler but also for stdlib; this is the only way to avoid endless discussions regarding whether something should belong to stdlib or nimble package (while allowing stdlib modules to depend on such modules); towards this end, we should make atlas install a dummy dependency as part of bootstrap to ensure this process works and keeps working (eg a lib/std/private/usesatlas.nim can be added which imports some atlas-installed package); once this is stable, actual dependencies can be used.

## questions for reviewer
`outfilecfg` will contain `--noNimblePath`; how can this work? won't that prevent users from using nimble installed packages?
```
############# begin Atlas config section ##########
--noNimblePath
--path:"../balls"
--path:"../grok"
...
############# end Atlas config section   ##########
```

what's the exact intention behind `--noNimblePath`, and can we use https://github.com/nim-lang/RFCs/issues/291 instead, which would be far more flexible?

### CODE REVIEW

File: tools/atlas/atlas.nim [hunk 1]
```diff
@@ -24,6 +36,8 @@ Command:
   search keyw keywB...  search for package that contains the given keywords
 
 Options:
+  --outFileCfg:file     (required) write cfg to `file`
+  --atlasDir:dir        (required) root of atlas packages
```
Comment (Araq) [inline tools/atlas/atlas.nim:40]: why are these required now? They should default to the cwd.

Comment (timotheecour) [reply tools/atlas/atlas.nim:40]: see linked issues ...

File: tools/atlas/atlas.nim [hunk 2]
```diff
@@ -435,12 +440,17 @@ proc main =
...
       else: writeHelp()
     of cmdEnd: assert false, "cannot happen"
-
-  while c.workspace.len > 0 and dirExists(c.workspace / ".git"):
-    c.workspace = c.workspace.parentDir()
-
+  if c.workspace.len == 0:
+    error "--atlasDir:dir must be provided" # refs #18750
+  if c.outfilecfg.len == 0:
+    error "--outFileCfg:file must be provided" # refs #18751
```
Comment (Araq) [inline tools/atlas/atlas.nim:453]: No, make it work according to its docs, which you apparently never read but then do so.

Comment (timotheecour) [reply tools/atlas/atlas.nim:453]: indeed, i did miss tools/atlas/atlas.md ; it's easy to miss though since it's not well referenced, 
 i had searched docs in a few places (including https://nim-lang.github.io/Nim/tools.html, and just assumed all there was the small comment in tools/atlas/atlas.nim)

Comment (timotheecour) [reply tools/atlas/atlas.nim:453]: even after reading the docs I find the pre-existing behavior (via a workspace) to be the wrong design, as it doesn't work well with existing user setups (like the 1 i had which triggered https://github.com/nim-lang/Nim/issues/18750)

it's likely others will have the same problem; the workspace, if hard-coded, should be contained within the source (in-tree, not out-of-tree), that way you don't mess with external code and still allow multiple nim repos (or nimble package clones) to peacefully co-exist in same dir without interference.

Comment (Araq) [reply tools/atlas/atlas.nim:453]: Make it work as I documented it please, it works well for me and other use cases can be supported via command line overrides.

**Comment by Araq** [comment]
It's unclear what this PR does and it needs to be redone.

1. Read the documentation.
2. Fix real bugs.
3. or - do a real refactoring adding some value that don't confuse bugs/assertions with decent error handling.

**Comment by timotheecour** [comment]
> Read the documentation.

Where was the RFC for atlas? Please write one so the design decisions can be discussed properly instead of baked in as of the PR that introduced it. I'm not necessarily questioning the utility of the tool (depending on the direction it's headed), I'm questioning the specific design decision such as writing out-of-tree which breaks existing use cases, overwrites unrelated user files and goes against best practices.

Which projects use atlas? AFAIK this was the 1st attempt at using it, it doesn't look promising when the issues and PR I've opened are being closed (eg https://github.com/nim-lang/Nim/issues/18751#issuecomment-907827845, https://github.com/nim-lang/Nim/issues/18754#issuecomment-907058360, etc), the problems are still there for anyone brave enough to try to use it.

The RFC should in particular cover concrete use cases:
* P1 how it'd help with stdlib/compiler/tools depend on a 3rd party package; in particular those need reproducible builds so locking down hashes of transitive dependencies is need
* P2 when to use nimble vs atlas
* P3 how will it work on projects that already depend on nimble, given that it adds `--noNimblePath`; in particular, how to ensure the (transitive) dependencies (say `pkg/regex`) remain local and don't interfere with user code that also happens to depend on `pkg/regex` (possibly with a different, incompatible version)

## P3
Link to relevant discussion: https://discord.com/channels/371759389889003530/768367394547957761/880738592056565790 where i suggested `--moduleoverride` (https://github.com/nim-lang/Nim/pull/18496) as a simple (IMO elegant) solution to P3, which would allow this:
```nim
# atlas adds this to $nim/config/nim.cfg: --moduleoverride:pkg:../lib/std/atlas:std
```
which would cause `import pkg/regex` (and the code inside it referring to transitive packages, eg `import pkg/unicodedb`) to resolve to `import $nim/lib/std/atlas/regex`, for all modules within a specified  prefix (eg `std` to make this apply to stdlib modules).

In particular, this wouldn't need `--noNimblePath`, it wouldn't expose the privately imported package to client code, and wouldn't interfere with client code that uses pkg/regex (possibly different version)

> PMs (Nimble included) should be able to import parse_requires.nim for a common ground.

this was advertised as a major point, except that atlas doesn't understand most of the requires syntax (eg <= and >= are confusd) and is incompatible with how nimble does it, see issues I've filed (and the closed ones), refs https://github.com/nim-lang/Nim/issues?q=label%3Aatlas+

**Comment by Araq** [comment]
It's a tool for myself and doesn't need an RFC, much like your `nim_dbg` didn't need an RFC. We've clarified in the linked discussion that it should not be used for introducing deps into the compiler.

> atlas doesn't understand most of the requires syntax (eg <= and >= are confusd) and is incompatible with how nimble does it, see issues I've filed

I don't agree and I don't want to spend time on discussing it anymore.

**Comment by timotheecour** [comment]
well you wrote here: https://github.com/nim-lang/RFCs/issues/411#issuecomment-906653102

> So provide a Nimble package as the foundation for these other Nimble packages and use "atlas" to be able to use it in Nim's core. I wrote Atlas for a reason.

which is in direct contradiction

I wouldn't have wasted time if I had known it's a tool for yourself not suitable for adding dependencies to compiler/stdlib.

**Comment by Araq** [comment]
> I wouldn't have wasted time if I had known it's a tool for yourself not suitable for adding dependencies to compiler/stdlib.

Well found that out after the discussions, keep the order of events in mind.
